quick-protobuf/src/reader.rs

```diff
@@ -338,14 +343,28 @@ impl BytesReader {
     /// Reads bytes (Vec<u8>)
     #[inline]
     pub fn read_bytes<'a>(&mut self, bytes: &'a [u8]) -> Result<&'a [u8]> {
-        self.read_len_varint(bytes, |r, b| Ok(&b[r.start..r.end]))
+        self.read_len_varint(bytes, |r, b| {
+            b.get(r.start..r.end).ok_or_else(|| {
+                Error::Io(io::Error::new(
+                    io::ErrorKind::UnexpectedEof,
+                    "Cannot read next bytes",
+                ))
+            })
+        })
     }
 
     /// Reads string (String)
     #[inline]
     pub fn read_string<'a>(&mut self, bytes: &'a [u8]) -> Result<&'a str> {
         self.read_len_varint(bytes, |r, b| {
-            ::std::str::from_utf8(&b[r.start..r.end]).map_err(|e| e.into())
+            b.get(r.start..r.end)
+                .ok_or_else(|| {
```

Comment:
Is there a performance benefit to skipping the `and_then` and using `?` instead? (Sorry for the very tardy response, just getting back into things... had a baby a bit ago, it's been hectic.)

Something like:

```rust
    /// Reads string (String)
    #[inline]
    pub fn read_string<'a>(&mut self, bytes: &'a [u8]) -> Result<&'a str> {
        self.read_len_varint(bytes, |r, b| {
            ::std::str::from_utf8(b.get(r.start..r.end).ok_or_else(|| {
                Error::Io(io::Error::new(
                    io::ErrorKind::UnexpectedEof,
                    "Cannot read next bytes",
                ))
            })?)
            .map_err(|e| e.into())
        })
    }
```

Reply:
I think in both cases program's path will be the same:
 - try to get slice
 - if success then try to create `&str` from it

If `and_then()` called with `Err` variant it just returns this error.
Also, we talking about fail path, that IMHO, is very rare case in practice.

To be convinced we should benchmark both version on some real message handling.

Reply:
I would be extremely surprised if they were any difference between the two.

Reply:
Ya, I wasn't sure if closures allocated or not. It seems that they don't?

Reply:
no they don't